packages/expo-in-app-purchases/android/src/main/java/expo/modules/inapppurchases/BillingManager.java

```diff
@@ -0,0 +1,500 @@
+    }
+
+    /**
+     * Listener for the Billing client state to become connected
+     */
+    public interface ServiceConnectedListener {
+        void onServiceConnected(BillingResult resultCode);
+    }
+
+    public BillingManager(Activity activity) {
+        Log.d(TAG, "Creating Billing client.");
+        mActivity = activity;
+        mBillingUpdatesListener = new UpdateListener();
+        mBillingClient =
+                BillingClient
+                        .newBuilder(activity)
+                        .enablePendingPurchases()
+                        .setListener(this)
+                        .build();
+    }
+
+    public void startConnectionAndQueryHistory(final Promise promise) {
+        Log.d(TAG, "Starting setup.");
+
+        // Start setup. This is asynchronous and the specified listener will be called
+        // once setup completes.
+        // It also starts to report all the new purchases through onPurchasesUpdated() callback.
+        startServiceConnection(new Runnable() {
+            @Override
+            public void run() {
+                // Notifying the listener that billing client is ready
+                mBillingUpdatesListener.onBillingClientSetupFinished();
+                // IAB is fully set up. Now, let's get an inventory of stuff we own.
+                Log.d(TAG, "Setup successful. Querying inventory.");
+                queryPurchases(promise);
+            }
+        });
+    }
+
+    public void startServiceConnection(final Runnable executeOnSuccess) {
+        mBillingClient.startConnection(new BillingClientStateListener() {
+            @Override
+            public void onBillingSetupFinished(BillingResult billingResult) {
+                final int responseCode = billingResult.getResponseCode();
+                Log.d(TAG, "Setup finished. Response code: " + responseCode);
+
+                if (responseCode == BillingResponseCode.OK) {
+                    mIsServiceConnected = true;
+                    if (executeOnSuccess != null) {
+                        executeOnSuccess.run();
+                    }
+                }
+                mBillingClientResponseCode = responseCode;
+            }
+
+            @Override
+            public void onBillingServiceDisconnected() {
+                mIsServiceConnected = false;
+            }
+        });
+    }
+
+    /**
+     * Start a purchase or subscription replace flow
+     */
+    public void purchaseItemAsync(final String skuId, final String oldSku, final Promise promise) {
+        // Save promise to HashMap to resolve later
+        promises.put(PURCHASING_ITEM, promise);
+
+        // oldSku is for subscription replacements and may be null.
+        Runnable purchaseFlowRequest = new Runnable() {
+            @Override
+            public void run() {
+            Log.d(TAG, "Launching in-app purchase flow. Replace old SKU? " + (oldSku != null));
+
+            SkuDetails skuDetails = mSkuDetailsMap.get(skuId);
```

Comment:
I've noticed this method behaves differently in iOS and Android:

Say we call `purchaseItemAsync` with an SKU we **didn't** fetch earlier:
- On iOS, [we'll fetch the product associated with the SKU](https://github.com/expo/expo/blob/da03ffb627349fa20dc1c9382409231fb8ba6964/packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m#L91) and then launch the billing flow.
- On Android, we'll attempt to pull from the cached `mSkuDetailsMap` which will return null, and then probably crash with an NPE.

It's probably best to make both iOS and Android behave the same here - either both fetch the product by SKU and launch the billing flow, or neither.

Reply:
Good catch. Fixed in f5f7f9b